File: src/main/java/ru/lom/d/gui/ContactManageView.java
package ru.lom.d.gui;

import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.button.ButtonVariant;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.component.textfield.TextField;
import ru.lom.d.data.Contact;
import ru.lom.d.services.ContactService;
import ru.lom.d.services.ContactServiceImpl;

public class ContactManageView extends VerticalLayout {

    ContactService contactService = new ContactServiceImpl();

    private final RootView rootView;

    private boolean newContactFlag = false;

    TextField firstName = new TextField("First name");
    TextField lastName = new TextField("Last name");
    TextField email = new TextField("Email");
    TextField phoneNumber = new TextField("Phone number");
    VerticalLayout layout = new VerticalLayout();

    public ContactManageView(RootView rootView) {
        Button newContactBtn = new Button("Add new contact");
        Button saveBtn = new Button("Save");
        Button closeBtn = new Button("Close");

        this.rootView = rootView;

        saveBtn.addClickListener(e -> addNewContact());
        saveBtn.addThemeVariants(ButtonVariant.LUMO_PRIMARY);

        closeBtn.addClickListener(e -> layout.setVisible(false));

        newContactBtn.addClickListener(e -> showNewContactForm());
        add(newContactBtn);

        layout.add(new HorizontalLayout(firstName, lastName, email, phoneNumber));
        layout.add(saveBtn);
        layout.add(closeBtn);
        layout.setVisible(false);
        add(layout);
    }

    private void showNewContactForm() {
        firstName.setValue("");
        lastName.setValue("");
        email.setValue("");
        phoneNumber.setValue("");

        newContactFlag = true;

        layout.setVisible(true);
    }

    public void showContact(Contact contact) {
        if(contact != null) {
            firstName.setValue(contact.getFirstName());
            lastName.setValue(contact.getLastName());
            email.setValue(contact.getEmail());
            phoneNumber.setValue(contact.getPhoneNumber());

            newContactFlag = false;

            layout.setVisible(true);
        }
    }

    private void addNewContact() {
        if(newContactFlag) {
            Contact contact = new Contact(firstName.getValue(),
                    lastName.getValue(),
                    email.getValue(),
                    phoneNumber.getValue());
            contactService.addContact(contact);
        } else {
            Contact contact = rootView.getTableOfContact().getSelectedContact();
            contact.setFirstName(firstName.getValue());
            contact.setLastName(lastName.getValue());
            contact.setEmail(email.getValue());
            contact.setPhoneNumber(phoneNumber.getValue());
            contactService.updateContact(contact);
        }

        firstName.setValue("");
        lastName.setValue("");
        email.setValue("");
        phoneNumber.setValue("");

        layout.setVisible(false);
        rootView.getTableOfContact().fillGrid();
    }


}

File: src/main/java/ru/lom/d/gui/ContactManageView.kt
package ru.lom.d.gui

import com.vaadin.flow.component.button.Button
import com.vaadin.flow.component.button.ButtonVariant
import com.vaadin.flow.component.orderedlayout.HorizontalLayout
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.component.textfield.TextField
import ru.lom.d.data.Contact
import ru.lom.d.services.ContactService
import ru.lom.d.services.ContactServiceImpl


class ContactManageView(val rootView: RootView) : VerticalLayout() {
    var contactService: ContactService = ContactServiceImpl()
    private var newContactFlag = false
    var firstName = TextField("First name")
    var lastName = TextField("Last name")
    var email = TextField("Email")
    var phoneNumber = TextField("Phone number")
    var layout = VerticalLayout()

    init {
        val newContactBtn = Button("Add new contact")
        val saveBtn = Button("Save")
        val closeBtn = Button("Close")
        saveBtn.addClickListener { addNewContact() }
        saveBtn.addThemeVariants(ButtonVariant.LUMO_PRIMARY)
        closeBtn.addClickListener { layout.isVisible = false }
        newContactBtn.addClickListener { showNewContactForm() }
        add(newContactBtn)
        layout.add(HorizontalLayout(firstName, lastName, email, phoneNumber))
        layout.add(saveBtn)
        layout.add(closeBtn)
        layout.isVisible = false
        add(layout)
    }

    private fun showNewContactForm() {
        firstName.value = ""
        lastName.value = ""
        email.value = ""
        phoneNumber.value = ""
        newContactFlag = true
        layout.isVisible = true
    }

    fun showContact(contact: Contact?) {
        if (contact != null) {
            firstName.value = contact.firstName
            lastName.value = contact.lastName
            email.value = contact.email
            phoneNumber.value = contact.phoneNumber
            newContactFlag = false
            layout.isVisible = true
        }
    }

    private fun addNewContact() {
        if (newContactFlag) {
            val contact = Contact(firstName.value,
                    lastName.value,
                    email.value,
                    phoneNumber.value)
            contactService.addContact(contact)
        } else {
            val contact = rootView.tableOfContact.getSelectedContact()!!
            contact.firstName = firstName.value
            contact.lastName = lastName.value
            contact.email = email.value
            contact.phoneNumber = phoneNumber.value
            contactService.updateContact(contact)
        }
        firstName.value = ""
        lastName.value = ""
        email.value = ""
        phoneNumber.value = ""
        layout.isVisible = false
        rootView.tableOfContact.fillGrid()
    }


}
